src/dispatch/batch_builder.rs

```diff
@@ -0,0 +1,150 @@
+use crate::{
+    dispatch::{Dispatcher, DispatcherBuilder},
+    world::ResourceId,
+    Accessor, AccessorCow, DynamicSystemData, RunningTime, System, SystemData, World,
+};
+
+/// The `BatchBuilder` is responsible for the creation of the Batch.
+///
+/// The `Batch` is a Systam that have the capabilities to contain another dispatcher
+/// Thanks to this characteristic is possible to customize its dispatching
+///
+/// The BatchBuilder accept the dispatcher builder as parameter, and the type of the
+/// System that will drive the execution of the internal dispatcher.
+///
+/// Note that depending on the dependencies of the SubSystems the Batch
+/// can run in parallel with other Systems.
+/// In addition the Sub Systems can run in parallel within the Batch.
+#[derive(Default)]
+pub struct BatchBuilder<'a, 'b> {
+    dispatcher_builder: DispatcherBuilder<'a, 'b>,
+}
+
+impl<'a, 'b> BatchBuilder<'a, 'b> {
+    /// Create a Batch using a dispatcher builder
+    pub fn new(dispatcher_builder: DispatcherBuilder<'a, 'b>) -> Self {
+        BatchBuilder { dispatcher_builder }
+    }
+}
+
+impl<'a, 'b> BatchBuilder<'a, 'b> {
+    /// Build the BatchSystem that will dispatch the inner dispatcher.
+    ///
+    /// Add the returned system to any dispatcher.
+    pub fn build<T>(self) -> T
+    where
+        T: System<'a> + BatchController<'a, 'b> + Send + 'a,
+    {
+        let mut reads = self.dispatcher_builder.stages_builder.fetch_all_reads();
+        reads.extend( <T::BatchSystemData as SystemData>::reads() );
+        reads.sort();
+        reads.dedup();
+
+        let mut writes = self.dispatcher_builder.stages_builder.fetch_all_writes();
+        writes.extend(<T::BatchSystemData as SystemData>::reads());
+        writes.sort();
+        writes.dedup();
+        
+        let accessor = BatchAccessor::new(reads, writes);
+        let dispatcher = self.dispatcher_builder.build();
+
+        T::create(accessor, dispatcher)
+    }
+}
+
+/// The BatchAccessor is responsible for the notification of the read and write resources
+/// of the subsystems
+pub struct BatchAccessor {
+    reads: Vec<ResourceId>,
+    writes: Vec<ResourceId>,
+}
+
+impl BatchAccessor {
+    /// Create the BatchAccessor
+    pub fn new(reads: Vec<ResourceId>, writes: Vec<ResourceId>) -> Self {
+        BatchAccessor { reads, writes }
+    }
+}
+
+impl Accessor for BatchAccessor {
+    fn try_new() -> Option<Self> {
+        None
+    }
+
+    fn reads(&self) -> Vec<ResourceId> {
+        self.reads.clone()
+    }
+
+    fn writes(&self) -> Vec<ResourceId> {
+        self.writes.clone()
+    }
+}
+
+/// The BatchUncheckedWorld wrap an instance of the world.
+/// It's safe to be used only in the context of the Batch.
+pub struct BatchUncheckedWorld<'a>(pub &'a World);
```

Comment:
Why is this field `pub`?

Reply:
This is pub because when you define the custom BatchControllerSystem you want to set it as SystemData